File: src/nw4j/wrapper/c/plusplus/Mat4Funcs.java
/*
 * Copyright (c) 2024, Miracle-Masterpiсe <mrmiraclemasterpiece@gmail.com or https://t.me/MiracleMasterpiece>. All rights reserved.
 * Use is subject to license terms.
 * 
 * 
 * 
 * 
 * 
 * 
 * 
 * 
 * */
package nw4j.wrapper.c.plusplus;

import static java.lang.foreign.ValueLayout.JAVA_FLOAT;
import static java.lang.foreign.ValueLayout.JAVA_LONG;

import java.lang.foreign.FunctionDescriptor;
import java.lang.foreign.Linker;
import java.lang.foreign.Linker.Option;
import java.lang.foreign.SymbolLookup;
import java.lang.invoke.MethodHandle;

import nw4j.helpers.Helpers;
import nw4j.helpers.NativeType;
import nw4j.wrapper.c.allocators.MemoryAccessor;
import nw4j.wrapper.c.pointers.FloatPointer;

/**
 * A class with methods for working with float arrays, 
 * which are a two-dimensional 4x4 matrix.
 * 
 * @since 0.1
 * @author miracle-masterpiece
 * */
public final class Mat4Funcs {

	private Mat4Funcs() {}

	private static final MethodHandle multMatrix;
	private static final MethodHandle setIdentity;
	private static final MethodHandle setScale;
	private static final MethodHandle setTranslate;
	private static final MethodHandle setXRotate;
	private static final MethodHandle setYRotate;
	private static final MethodHandle setZRotate;
	private static final MethodHandle setOrtho;
	private static final MethodHandle setFrustum;
	private static final MethodHandle setProjection;
	private static final MethodHandle transpose;
	private static final MethodHandle add;
	private static final MethodHandle sub;
	private static final MethodHandle setZero;
	private static final MethodHandle negate;
	private static final MethodHandle toVector;

	static {
		Helpers.loadlib("MathNativeFuncs");

		multMatrix 		= createCriticalMethodHandle("mult_matrix", 	FunctionDescriptor.ofVoid(JAVA_LONG, JAVA_LONG, JAVA_LONG));
		setIdentity 	= createCriticalMethodHandle("setIdentity", 	FunctionDescriptor.ofVoid(JAVA_LONG));
		setScale 		= createCriticalMethodHandle("setScale", 		FunctionDescriptor.ofVoid(JAVA_LONG, JAVA_FLOAT, JAVA_FLOAT, JAVA_FLOAT));
		setTranslate 	= createCriticalMethodHandle("setTranslate", 	FunctionDescriptor.ofVoid(JAVA_LONG, JAVA_FLOAT, JAVA_FLOAT, JAVA_FLOAT));
		setXRotate 		= createCriticalMethodHandle("setXRotate", 		FunctionDescriptor.ofVoid(JAVA_LONG, JAVA_FLOAT));
		setYRotate 		= createCriticalMethodHandle("setYRotate", 		FunctionDescriptor.ofVoid(JAVA_LONG, JAVA_FLOAT));
		setZRotate 		= createCriticalMethodHandle("setZRotate", 		FunctionDescriptor.ofVoid(JAVA_LONG, JAVA_FLOAT));
		setOrtho 		= createCriticalMethodHandle("setOrtho", 		FunctionDescriptor.ofVoid(JAVA_LONG, JAVA_FLOAT, JAVA_FLOAT, JAVA_FLOAT, JAVA_FLOAT, JAVA_FLOAT, JAVA_FLOAT));
		setFrustum 		= createCriticalMethodHandle("setFrustum", 		FunctionDescriptor.ofVoid(JAVA_LONG, JAVA_FLOAT, JAVA_FLOAT, JAVA_FLOAT, JAVA_FLOAT, JAVA_FLOAT, JAVA_FLOAT));
		setProjection 	= createCriticalMethodHandle("setProjection", 	FunctionDescriptor.ofVoid(JAVA_LONG, JAVA_FLOAT, JAVA_FLOAT, JAVA_FLOAT, JAVA_FLOAT));
		transpose 		= createCriticalMethodHandle("transpose", 		FunctionDescriptor.ofVoid(JAVA_LONG));
		add 			= createCriticalMethodHandle("add", 			FunctionDescriptor.ofVoid(JAVA_LONG, JAVA_LONG, JAVA_LONG));
		sub 			= createCriticalMethodHandle("sub", 			FunctionDescriptor.ofVoid(JAVA_LONG, JAVA_LONG, JAVA_LONG));		
		setZero 		= createCriticalMethodHandle("setZero", 		FunctionDescriptor.ofVoid(JAVA_LONG));
		negate 			= createCriticalMethodHandle("negate", 			FunctionDescriptor.ofVoid(JAVA_LONG));
		toVector		= createCriticalMethodHandle("mult_vector", 	FunctionDescriptor.ofVoid(JAVA_LONG, JAVA_LONG));
	}

	/**
	 * Multiplies a matrix by a vector and puts the result in the transmitted vector.
	 * @param mat4 	A float array that is a matrix4x4.
	 * @param vec4 	A float array that is a vector4.
	 * */
	public static @NativeType("float*") long toVector(@NativeType("float*") long mat4, @NativeType("float*") long vec4) {
		try{
			toVector.invoke(mat4, vec4);
		}catch (Throwable e) {throw new RuntimeException(e);}
		return vec4;
	}

	/**
	 * Multiplies a matrix by -1 and puts the result in the transmitted matrix.
	 *    / 0  1  2  3  \ 
	 * -1*| 5  6  7  8  |
	 *	  | 8  9  10 11 |
	 *	  \	12 13 14 15 /
	 *
	 * @param mat4 	Matrix that will be multiplied by -1.
	 * */
	public static @NativeType("float*") long negate(@NativeType("float*") long mat4) {
		try{
			negate.invoke(mat4);
		}catch (Throwable e) {throw new RuntimeException(e);}
		return mat4;
	}

	/**
	 * Sets all values in a matrix to 0.
	 * @param mat4 	A float array that is a 4x4 matrix.
	 * */
	public static @NativeType("float*") long setZero(@NativeType("float*") long mat4) {
		try{
			setZero.invoke(mat4);
		}catch (Throwable e) {throw new RuntimeException(e);}
		return mat4;
	}

	/**
	 * Subtracts matrix B from matrix A and puts the result into matrix result.
	 * @param A 		First matrix.
	 * @param B 		Second matrix.
	 * @param result 	Result matrix.
	 * */
	public static @NativeType("float*") long sub(@NativeType("float*") long A, @NativeType("float*") long B, @NativeType("float*") long result) {
		try{
			sub.invoke(A, B, result);
		}catch (Throwable e) {throw new RuntimeException(e);}
		return result;
	}

	/**
	 * Adds matrix A and B and puts the result in the result matrix.
	 * @param A 		First matrix.
	 * @param B 		Second matrix.
	 * @param result 	Result matrix.
	 * */
	public static @NativeType("float*") long add(@NativeType("float*") long a, @NativeType("float*") long b, @NativeType("float*") long result) {
		try{
			add.invoke(a, b, result);
		}catch (Throwable e) {throw new RuntimeException(e);}
		return result;
	}

	/**
	 * Transposes a matrix
	 *    
	 * Before transpose:
	 *    / 0  1  2  3  \ 
	 * 	  | 5  6  7  8  |
	 *	  | 8  9  10 11 |
	 *	  \	12 13 14 15 /
	 * 	  
	 * After transpose:
	 * 	  / 0  4  8  12 \ 
	 * 	  | 1  5  9  13 |
	 *	  | 2  6  10 14 |
	 *	  \	3  7  11 15 /
	 * 
	 * */
	public static @NativeType("float*") long transpose(@NativeType("float*") long mat4) {
		try{
			transpose.invoke(mat4);
		}catch (Throwable e) {throw new RuntimeException(e);}
		return mat4;
	}

	/**
	 * Sets matrix values as projection matrices.
	 * */
	public static @NativeType("float*") long projection(@NativeType("float*") long mat4, float fov, float aspect, float znear, float zfar) {
		try{
			setProjection.invoke(mat4, fov, aspect, znear, zfar);
		}catch (Throwable e) {throw new RuntimeException(e);}
		return mat4;
	}

	/**
	 * Sets matrix values as frustum matrices.
	 * */
	public static @NativeType("float*") long frustum(@NativeType("float*")long mat4, float left, float right, float bottom, float top, float znear, float zfar) {
		try{
			setFrustum.invoke(mat4, left, right, bottom, top, znear, zfar);
		}catch (Throwable e) {throw new RuntimeException(e);}
		return mat4;
	}

	/**
	 * Sets matrix values as ortho matrices.
	 * */
	public static @NativeType("float*") long ortho(@NativeType("float*")long mat4, float left, float right, float bottom, float top, float znear, float zfar) {
		try{
			setOrtho.invoke(mat4, left, right, bottom, top, znear, zfar);
		}catch (Throwable e) {throw new RuntimeException(e);}
		return mat4;
	}

	/**
	 * Sets matrix values as zRotate matrices.
	 * 
	 * @param mat4 	A float array that is a 4x4 matrix.
	 * @param angle Angle to rotate.
	 * */
	public static @NativeType("float*") long zRotate(@NativeType("float*") long mat4, float angle) {
		try{
			setZRotate.invoke(mat4, angle);
		}catch (Throwable e) {throw new RuntimeException(e);}
		return mat4;
	}

	/**
	 * Sets matrix values as yRotate matrices.
	 * 
	 * @param mat4 	A float array that is a 4x4 matrix.
	 * @param angle Angle to rotate.
	 * */
	public static @NativeType("float*") long yRotate(@NativeType("float*") long mat4, float angle) {
		try{
			setYRotate.invoke(mat4, angle);
		}catch (Throwable e) {throw new RuntimeException(e);}
		return mat4;
	}

	/**
	 * Sets matrix values as xRotate matrices.
	 * 
	 * @param mat4 	A float array that is a 4x4 matrix.
	 * @param angle Angle to rotate.
	 * */
	public static @NativeType("float*") long xRotate(@NativeType("float*") long mat4, float angle) {
		try{
			setXRotate.invoke(mat4, angle);
		}catch (Throwable e) {throw new RuntimeException(e);}
		return mat4;
	}

	/**
	 * Sets matrix values as translate matrices.
	 * 
	 * @param mat4 	A float array that is a 4x4 matrix.
	 * @param x 	Translate x axis.
	 * @param y 	Translate y axis.
	 * @param z 	Translate z axis.
	 * */
	public static @NativeType("float*") long translate(@NativeType("float*") long mat4, float x, float y, float z) {
		try{
			setTranslate.invoke(mat4, x, y, z);
		}catch (Throwable e) {throw new RuntimeException(e);}
		return mat4;
	}

	/**
	 * Sets matrix values as scale matrices.
	 * 
	 * @param mat4 	A float array that is a 4x4 matrix.
	 * @param x 	Scale x axis.
	 * @param y 	Scale y axis.
	 * @param z 	Scale z axis.
	 * */
	public static @NativeType("float*") long scale(@NativeType("float*") long mat4, float x, float y, float z) {
		try{
			setScale.invoke(mat4, x, y, z);
		}catch (Throwable e) {throw new RuntimeException(e);}
		return mat4;
	}

	/**
	 * Multiplies matrix A by matrix B and puts the result in the result matrix.
	 * 
	 * Note: 
	 * The result can also be set to the matrix that participated in the operation, 
	 * that is, to matrix A or B, you just need to pass the result matrix pointer to matrix A or B.
	 * 
	 * @param A 		A float array that is a 4x4 matrix.
	 * @param B 		A float array that is a 4x4 matrix.
	 * @param result 	A float array that is a 4x4 matrix. Where store to result.
	 * */
	public static @NativeType("float*") long mult(@NativeType("float*") long A, @NativeType("float*") long B, @NativeType("float*") long result) {
		try{
			multMatrix.invoke(A, B, result);
		}catch (Throwable e) {throw new RuntimeException(e);}
		return result;
	}

	/**
	 * Sets the matrix to be the identity matrix.
	 * @param mat4 		A float array that is a 4x4 matrix.
	 * */
	public static @NativeType("float*") long loadIdentity(@NativeType("float*") long mat4) {
		try{
			setIdentity.invoke(mat4);
		}catch (Throwable e) {throw new RuntimeException(e);}
		return mat4;
	}

	//gen critical methodHandle
	private static MethodHandle createCriticalMethodHandle(String name, FunctionDescriptor desc) {
		final Linker nativeLinker 	= Linker.nativeLinker();
		final Option critical 		= Option.critical(false);
		MethodHandle handle = null;
		try{
			handle = nativeLinker.downcallHandle(SymbolLookup.loaderLookup().find(name).orElseThrow(), desc, critical);
		}catch(Exception e) {
			throw e;
		}
		return handle;
	}

	//print array as matrix4x4
	public static void print(FloatPointer ptr) {
		for (int y = 0; y < 4; ++y) {
			for (int x = 0; x < 4; ++x) {
				int index = x + (y * 4);
				System.out.print(ptr.get(index) + " ");
			}
			System.out.println();
		}
	}

	//print array as matrix4x4
	public static void print(long ptr) {
		for (int y = 0; y < 4; ++y) {
			for (int x = 0; x < 4; ++x) {
				int index = x + (y * 4);
				System.out.print(MemoryAccessor.getFloat(ptr+(index*4)) + " ");
			}
			System.out.println();
		}
	}
}

File: native/src/Math/include/matrix/mat4funcs.c
/*
 * Copyright (c) 2024, Miracle-Masterpiсe <mrmiraclemasterpiece@gmail.com or https://t.me/MiracleMasterpiece>. All rights reserved.
 * Use is subject to license terms.
 * 
 * 
 * 
 * 
 * 
 * 
 * 
 * 
 * */
#ifdef USE_INTRINSIC
	#include <xmmintrin.h>
#endif

#include <stdio.h>
#include <string.h>
#include <inttypes.h>

	void print(jfloat* mat4) {
		for (int y = 0; y < 4; ++y) {
			for (int x = 0; x < 4; ++x) {
				int index = x + (y * 4);
				printf("%f ", mat4[index]);
			}
			printf("\n");
		}
		printf("\n");
	}	

	void printVec(jfloat* vec){
		for (int i = 0; i < 4; ++i){
			printf("%f ", vec[i]);
		}
		printf("\n");
	}

	void mult_matrix(jfloat* A, jfloat* B, jfloat* C) {

		#ifdef USE_INTRINSIC
			//This code is copied from stackoverflow xD		
			const __m128 row1 = _mm_load_ps(&B[0]);
    		const __m128 row2 = _mm_load_ps(&B[4]);
    		const __m128 row3 = _mm_load_ps(&B[8]);
    		const __m128 row4 = _mm_load_ps(&B[12]);
    		for(int i = 0; i < 4; i++) {
       			__m128 brod1 = 	_mm_set1_ps(A[4*i + 0]);
        		__m128 brod2 = 	_mm_set1_ps(A[4*i + 1]);
        		__m128 brod3 = 	_mm_set1_ps(A[4*i + 2]);
        		__m128 brod4 = 	_mm_set1_ps(A[4*i + 3]);
        	
        		__m128 row   = 	_mm_add_ps(_mm_add_ps(
                       			_mm_mul_ps(brod1, row1),
                       			_mm_mul_ps(brod2, row2)),
                       			_mm_add_ps(
                       			_mm_mul_ps(brod3, row3),
                       			_mm_mul_ps(brod4, row4)));
        		_mm_store_ps(&C[4*i], row);
    	}
		#else
			float tmp[16];			
			for (int i = 0; i < 16; i+=4) {
				for (int j = 0; j < 4; j++) {
					tmp[i + j] = (A[i] * B[j]) + (A[i + 1] * B[j + 4]) + (A[i + 2] * B[j + 8])  + (A[i + 3] * B[j + 12]);		
				}
			}
			memcpy(tmp, C, sizeof(jfloat) * 16);
		#endif
	}

	void mult_vector(jfloat* mat4, jfloat* vec4){
		/**
		 * 0		0  1  2  3 			0
		 * 1		4  5  6  7			1
		 * 2		8  9  10 11			2
		 * 3		12 13 14 15			3
		 */
		#ifdef USE_INTRINSIC
		__m128 vec = _mm_load_ps(vec4);
			for (int i = 0; i < 4; ++i){
				__m128 mat = _mm_load_ps(&mat4[i * 4]);
				__m128 res = _mm_mul_ps(mat, vec);
				vec4[i] = res[0] + res[1] + res[2] + res[3];
			}
		#else
			jfloat result[4] = {0, 0, 0, 0};
			for (int j = 0; j < 4; ++j){
		 		for (int i = 0; i < 4; ++i){
		 			result[j] += mat4[j * 4 + i] * vec4[i];
		 		}				
		 	}
			memcpy(vec4, result, (sizeof(jfloat) * 4));
		#endif
	}

	void setIdentity(jfloat* mat4){
		mat4[0]  = 1; 	mat4[1]  = 0; 	mat4[2]  = 0; 	mat4[3]  = 0;
		mat4[4]  = 0; 	mat4[5]  = 1; 	mat4[6]  = 0; 	mat4[7]  = 0;
		mat4[8]  = 0; 	mat4[9]  = 0; 	mat4[10] = 1; 	mat4[11] = 0;
		mat4[12] = 0; 	mat4[13] = 0; 	mat4[14] = 0; 	mat4[15] = 1;
	}
	
	void setScale(jfloat* mat4, jfloat x, jfloat y, jfloat z){
		mat4[0]  = x; 	mat4[1]  = 0; 	mat4[2]  = 0; 	mat4[3]  = 0;
		mat4[4]  = 0; 	mat4[5]  = y; 	mat4[6]  = 0; 	mat4[7]  = 0;
		mat4[8]  = 0; 	mat4[9]  = 0; 	mat4[10] = z; 	mat4[11] = 0;
		mat4[12] = 0; 	mat4[13] = 0; 	mat4[14] = 0; 	mat4[15] = 1;
	}
	
	void setTranslate(jfloat* mat4, jfloat x, jfloat y, jfloat z){
		mat4[0]  = 1; 	mat4[1]  = 0; 	mat4[2]  = 0; 	mat4[3]  = x;
		mat4[4]  = 0; 	mat4[5]  = 1; 	mat4[6]  = 0; 	mat4[7]  = y;
		mat4[8]  = 0; 	mat4[9]  = 0; 	mat4[10] = 1; 	mat4[11] = z;
		mat4[12] = 0; 	mat4[13] = 0; 	mat4[14] = 0; 	mat4[15] = 1;
	}
	
	void setXRotate(jfloat* mat4, jfloat angle){
		const jfloat rad = angle * TO_RAD;
		const jfloat sin_ = sinf(rad);
		const jfloat cos_ = cosf(rad); 	
		
		mat4[0]  = 1; 	mat4[1]  = 0; 		mat4[2]  = 0; 		mat4[3]  = 0;
		mat4[4]  = 0; 	mat4[5]  = cos_; 	mat4[6]  = -sin_; 	mat4[7]  = 0;
		mat4[8]  = 0; 	mat4[9]  = sin_; 	mat4[10] =  cos_; 	mat4[11] = 0;
		mat4[12] = 0; 	mat4[13] = 0; 		mat4[14] = 0; 		mat4[15] = 1;
	}

	void setYRotate(jfloat* mat4, jfloat angle){
		const jfloat rad = angle * TO_RAD;
		const jfloat sin_ = sinf(rad);
		const jfloat cos_ = cosf(rad); 	
		
		mat4[0]  = cos_; 	mat4[1]  = 0; 	mat4[2]  = -sin_; 	mat4[3]  = 0;
		mat4[4]  = 0; 		mat4[5]  = 1; 	mat4[6]  = 0; 		mat4[7]  = 0;
		mat4[8]  = sin_; 	mat4[9]  = 0; 	mat4[10] = cos_; 	mat4[11] = 0;
		mat4[12] = 0; 		mat4[13] = 0; 	mat4[14] = 0; 		mat4[15] = 1;
	}

	void setZRotate(jfloat* mat4, jfloat angle){
		const jfloat rad = angle * TO_RAD;
		const jfloat sin_ = sinf(rad);
		const jfloat cos_ = cosf(rad); 	
		
		mat4[0]  = cos_; 	mat4[1]  = -sin_; 	mat4[2]  = 0; 	mat4[3]  = 0;
		mat4[4]  = sin_; 	mat4[5]  = cos_; 	mat4[6]  = 0; 	mat4[7]  = 0;
		mat4[8]  = 0; 		mat4[9]  = 0; 		mat4[10] = 1; 	mat4[11] = 0;
		mat4[12] = 0; 		mat4[13] = 0; 		mat4[14] = 0; 	mat4[15] = 1;
	}
	
	void setOrtho(jfloat* mat4, jfloat left, jfloat right, jfloat bottom, jfloat top, jfloat znear, jfloat zfar) {
		const jfloat m0 =  2 / (right - left);
		const jfloat m3 =  -((right+left)/(right - left));
		const jfloat m5 =  2 / (top - bottom);
		const jfloat m7 =  -((top+bottom)/(top - bottom));
		const jfloat m10 = 2/(zfar - znear);
		const jfloat m11 = -((zfar+znear)/(zfar - znear));

		/**
		 * 
		 *	-m0,   0,    0,    m3,
		 *	 0,   -m5,   0,    m7,
		 *	 0,    0,   -m10,  m11,
		 *	 0,    0,    0,    1
		 * 
		 * */	
		mat4[0]  = m0; 	mat4[1]  = 0; 	mat4[2]   = 0; 	  mat4[3]  = m3;
		mat4[4]  = 0;  	mat4[5]  = m5;	mat4[6]   = 0;	  mat4[7]  = m7;
		mat4[8]  = 0;	mat4[9]  = 0;	mat4[10]  = m10;  mat4[11] = m11;
		mat4[12] = 0;   mat4[13] = 0;   mat4[14]  = 0;	  mat4[15] = 1;
	}
	
	void setFrustum(jfloat* mat4, jfloat left, jfloat right, jfloat bottom, jfloat top, jfloat znear, jfloat zfar) {
		const jfloat rightMinusLeft 	= right - left;
		const jfloat topMinusBottom 	= top - bottom;
		const jfloat zfarMinusZnear 	= zfar - znear;
		const jfloat twoMultZnear 		= 2 * znear;
		
		const jfloat m0 	= (twoMultZnear) 	 / (rightMinusLeft);
		const jfloat m2 	= (right+left) 	 	 / (rightMinusLeft);
		const jfloat m5 	= (twoMultZnear) 	 / (topMinusBottom);
		const jfloat m6 	= (top + bottom) 	 / (topMinusBottom);
		const jfloat m10 	= (zfar + znear) 	 / (zfarMinusZnear);
		const jfloat m11 	= (2 * zfar * znear) / (zfarMinusZnear);
		
		mat4[0]  = m0;	mat4[1]  = 0;	mat4[2]  = m2;	mat4[3]  = 0;
		mat4[4]  = 0;	mat4[5]  = m5;	mat4[6]	 = m6;	mat4[7]  = 0;
		mat4[8]  = 0;	mat4[9]  = 0;	mat4[10] =-m10;	mat4[11] =-m11;
		mat4[12] = 0;	mat4[13] = 0;	mat4[14] =-1;	mat4[15] = 0;
	}
	
	void setProjection(jfloat* mat4, jfloat fov, jfloat aspect, jfloat znear, jfloat zfar){
		const jfloat radiands = (fov / 2) * TO_RAD;
		const jfloat sin_ = sin(radiands);
		const jfloat ctg =  cos(radiands) / sin_;
		const jfloat zEpsilon = zfar - znear;

		/**
		 * 0  1  2  3
		 * 4  5  6  7      
		 * 8  9  10 11       
		 * 12 13 14 15		
		 * */
		mat4[0] =   ctg / aspect;
		mat4[5] =   ctg;
		mat4[10] = -(zfar + znear) / zEpsilon;
		if (0) { //transpose
			mat4[14] = -1;
			mat4[11] =  (-2 * znear * zfar) / zEpsilon;
			mat4[15] =  0;
		}else {
			mat4[11] = -1;
			mat4[14] =  (-2 * znear * zfar) / zEpsilon;
			mat4[15] =  0;
		}
	}
	
	void transpose(jfloat* const mat4) {
		for (uint32_t y = 0; y < 3; ++y) {
			for (uint32_t x = y; x < 4; ++x) {
				const uint32_t in 	= x + (y << 2 /*a.k.a * 4*/);
				const uint32_t out 	= y + (x << 2 /*a.k.a * 4*/);
				const jfloat tmp1 = mat4[in];
				const jfloat tmp2 = mat4[out];
				mat4[in] 	= tmp2;
				mat4[out] 	= tmp1;				
			}
		}
	}
	void add(jfloat* const A, jfloat* const B, jfloat* const result){
		#ifdef USE_INTRINSIC
			for (uint32_t i = 0; i < 4; ++i){
				const uint32_t index = i << 2;/*a.k.a * 4*/
				__m128 a = _mm_load_ps(&A[index]);
    			__m128 b = _mm_load_ps(&B[index]);
				__m128 c = _mm_add_ps(a, b);	
				_mm_store_ps(&result[index], c);
			}
		#else
			for (uint32_t i = 0; i < 4; ++i){
				result[i] = A[i] + B[i];
			}
		#endif
	}

	void sub(jfloat* const A, jfloat* const B, jfloat* const result){
		#ifdef USE_INTRINSIC
			for (uint32_t i = 0; i < 4; ++i){
			const uint32_t index = i << 2;/*a.k.a * 4*/
			__m128 a = _mm_load_ps(&A[index]);
    		__m128 b = _mm_load_ps(&B[index]);
			__m128 c = _mm_sub_ps(a, b);	
			_mm_store_ps(&result[index], c);
		}
		#else
			for (uint32_t i = 0; i < 4; ++i){
				result[i] = A[i] - B[i];
			}
		#endif
	}

	///=============== new ========================///
	void setZero(jfloat* const mat4){
		memset(mat4, 0, sizeof(jfloat) * 4);
	}

	void negate(jfloat* const mat4){
		#ifdef USE_INTRINSIC
			const float A[] = {-1, -1, -1, -1};
			__m128 oneNegative = _mm_load_ps(A); 
			for (int i = 0; i < 4; ++i){
				__m128 row 		= _mm_load_ps(&mat4[i * 4]); 
				__m128 result 	= _mm_mul_ps(row, oneNegative);
				_mm_store_ps(&mat4[i*4], result);
			}
		#else
			for (int i = 0; i < 16; ++i){
				mat4[i] *= -1;
			}
		#endif
	}
